auth.go

```diff
@@ -0,0 +1,160 @@
+package main
+
+import (
+	"bufio"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"golang.org/x/crypto/ssh/terminal"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"syscall"
+)
+
+const (
+	https        = "https"
+	accessToken = "access_token"
+)
+
+type tokenRequest struct {
+	User     string
+	Password string
+	Realm    *string
+	Scopes   []string
+}
+
+// Authentication abstracts the HTTP request authentication, allowing to forward the credentials with the request
+type Authentication interface {
+	Enrich(*http.Request)
+}
+
+// Authenticator abstract the authentication process
+type Authenticator interface {
+	Authenticate() (Authentication, error)
+}
+
+// BearerTokenAuthentication the OAuth2 Bearer token authentication
+type BearerTokenAuthentication struct {
+	token *string
+}
+
+// Enrich allows to add the authentication details to the request
+func (auth BearerTokenAuthentication) Enrich(req *http.Request) {
+	if auth.token != nil {
+		req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", auth.token))
+	}
+}
+
+// OAuth2Authenticator hides all of the details of authenticating against authorization server
+type OAuth2Authenticator struct {
+	url string
+}
+
+// NewOAuth2Authenticator instantiates new OAuth2Authenticator
+func NewOAuth2Authenticator(url string) Authenticator {
+	return OAuth2Authenticator{url: url}
+}
+
+// Authenticate against the configured OAuth2 authorization server
+func (auth OAuth2Authenticator) Authenticate() (Authentication, error) {
+	// TODO implement credentials
+
+	reader := bufio.NewReader(os.Stdin)
+	fmt.Print("Enter user name: ")
+	user, err := reader.ReadString('\n')
+	if err != nil {
+		return nil, err
+	}
+	fmt.Print("Enter password: ")
+	password, err := terminal.ReadPassword(int(syscall.Stdin))
+	if err != nil {
+		return nil, err
+	}
+	fmt.Println()
```

Comment:
I think reading username/password should be handled in the cli part of the code and then passed to the auth code.